lib/galaxy/datatypes/data.py

```diff
@@ -34,17 +38,26 @@
     FILENAME_VALID_CHARS,
     inflector,
     iter_start_of_line,
-    smart_str,
     unicodify,
 )
 from galaxy.util.bunch import Bunch
+from galaxy.util.markdown import MarkdownFormatHelpers
 from galaxy.util.sanitize_html import sanitize_html
 from galaxy.util.zipstream import ZipstreamWrapper
 from . import dataproviders as p_dataproviders
 from . import metadata
 
 if TYPE_CHECKING:
-    from galaxy.model import DatasetInstance
+    from galaxy.datatypes.display_applications.application import DisplayApplication
+    from galaxy.datatypes.registry import Registry
+    from galaxy.model import (
+        DatasetInstance,
+        History,
+        HistoryDatasetAssociation,
+        HistoryDatasetCollectionAssociation,
+    )
+    from galaxy.tools import Tool
```

Comment:
I'm -0 on all these forward references. We shouldn't be importing code we don't expect to be available in the packages. This includes GalaxyWebTransaction and Tool. The type checking will fail in those cases as we tighten up our mypy settings to disallow unknown imports.

-0 and not -1 because I will acknowledge fixes I would find palatable are challenging tasks of unwinding these dependencies or declaring protocols for how we consume app and trans in this package.

Reply:
My primary goal here is to add type annotations for dataset-like parameters (DatasetInstances + not DatasetInstances that are also "datasets" of another sort). So I can do without most other forward references. I'll drop GalaxyWebTransaction and Tool (I'll check the data package for imports to avoid whatever else I can). But yes, declaring protocols for those is a project of its own. If you don't mind I'll keep the dataset-related imports for now - I suppose we can replace them once we get to defining pydantic models for them (which I'd rather not tackle here).